xiilib/flask/charm.py

```diff
@@ -1,230 +1,113 @@
-                getattr(self, f"_on_{database}_database_relation_broken"),
-            )
+        Returns:
+            The string converted to uppercase.
+        """
+        return value.upper()
 
-    def _on_config_changed(self, _event: ops.EventBase) -> None:
-        """Configure the flask pebble service layer.
 
-        Args:
-            _event: the config-changed event that triggers this callback function.
-        """
-        self._restart_flask()
+class Charm(GunicornBase):  # pylint: disable=too-many-instance-attributes
+    """Flask Charm service."""
 
-    def _on_rotate_secret_key_action(self, event: ops.ActionEvent) -> None:
-        """Handle the rotate-secret-key action.
+    def __init__(self, framework: ops.Framework) -> None:
+        """Initialize the Flask charm.
 
         Args:
-            event: the action event that trigger this callback.
+            framework: operator framework.
         """
-        if not self.unit.is_leader():
-            event.fail("only leader unit can rotate secret key")
-            return
-        if not self._secret_storage.is_initialized:
-            event.fail("flask charm is still initializing")
-            return
-        self._secret_storage.reset_flask_secret_key()
-        event.set_results({"status": "success"})
-        self._restart_flask()
-
-    def _on_secret_storage_relation_changed(self, _event: ops.RelationEvent) -> None:
-        """Handle the secret-storage-relation-changed event.
+        super().__init__(framework=framework, wsgi_framework="flask")
+        self.framework.observe(self.on.flask_app_pebble_ready, self._on_flask_app_pebble_ready)
 
-        Args:
-            _event: the action event that triggers this callback.
-        """
-        self._restart_flask()
+    def get_wsgi_config(self) -> BaseModel:
+        """Return Flask framework related configurations.
 
-    def _update_app_and_unit_status(self, status: ops.StatusBase) -> None:
-        """Update the application and unit status.
+        Returns:
+             Flask framework related configurations.
 
-        Args:
-            status: the desired application and unit status.
+        Raises:
+            CharmConfigInvalidError: if charm config is not valid.
         """
-        self.unit.status = status
-        if self.unit.is_leader():
-            self.app.status = status
-
-    def _restart_flask(self) -> None:
-        """Restart or start the flask service if not started with the latest configuration."""
+        flask_config = {
+            k.removeprefix("flask-").replace("-", "_"): v
+            for k, v in self.config.items()
+            if k.startswith("flask-")
+        }
         try:
-            self._flask_app.restart()
-            self._update_app_and_unit_status(ops.ActiveStatus())
-        except CharmConfigInvalidError as exc:
-            self._update_app_and_unit_status(ops.BlockedStatus(exc.msg))
+            return FlaskConfig(**flask_config)  # type: ignore
```

Comment:
You can use `model_validate` from pydantic instead that avoids `**` unpacking and `type: ignore`

Reply:
Gotcha, let me take a look at `model_validate` function

Reply:
I have check the `model_validate` function and it doesn't seem to applicable here as we need the model as the return value and it seems that `model_validate` doesn't return the model.